src/util/mpsc.rs

```diff
@@ -0,0 +1,268 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*!
+
+This module provides an implementation of mpsc channel based on
+crossbeam_channel. Comparing to the crossbeam_channel, this implementation
+supports closed detection and try operations.
+
+*/
+
+use crossbeam_channel;
+use std::sync::atomic::{AtomicIsize, Ordering};
+use std::sync::mpsc::{RecvError, RecvTimeoutError, SendError, TryRecvError, TrySendError};
+use std::sync::Arc;
+use std::time::Duration;
+
+struct ReplicaCount {
+    sender: AtomicIsize,
+    receiver: AtomicIsize,
+}
+
+impl ReplicaCount {
+    #[inline]
+    fn new() -> ReplicaCount {
+        ReplicaCount {
+            sender: AtomicIsize::new(1),
+            receiver: AtomicIsize::new(1),
+        }
+    }
+
+    #[inline]
+    fn is_receiver_closed(&self) -> bool {
+        self.receiver.load(Ordering::Acquire) == 0
+    }
+
+    #[inline]
+    fn is_sender_closed(&self) -> bool {
+        self.sender.load(Ordering::Acquire) == 0
+    }
+}
+
+pub struct Sender<T> {
+    sender: crossbeam_channel::Sender<T>,
+    replica_cnt: Arc<ReplicaCount>,
+}
+
+impl<T> Clone for Sender<T> {
+    #[inline]
+    fn clone(&self) -> Sender<T> {
+        self.replica_cnt.sender.fetch_add(1, Ordering::AcqRel);
+        Sender {
+            sender: self.sender.clone(),
+            replica_cnt: self.replica_cnt.clone(),
+        }
+    }
+}
+
+impl<T> Drop for Sender<T> {
+    #[inline]
+    fn drop(&mut self) {
+        self.replica_cnt.sender.fetch_add(-1, Ordering::AcqRel);
+    }
+}
+
+pub struct Receiver<T> {
+    receiver: crossbeam_channel::Receiver<T>,
+    replica_cnt: Arc<ReplicaCount>,
+}
+
+impl<T> Sender<T> {
+    #[inline]
+    pub fn send(&self, t: T) -> Result<(), SendError<T>> {
+        if !self.replica_cnt.is_receiver_closed() {
+            self.sender.send(t);
+            return Ok(());
+        }
+        Err(SendError(t))
+    }
+
+    #[inline]
+    pub fn try_send(&self, t: T) -> Result<(), TrySendError<T>> {
+        if !self.replica_cnt.is_receiver_closed() {
+            return select! {
```

Comment:
Please add some comments about the macro, and where it's from.

Reply:
It's quite common to use the macro just like `try_ready` from futures, it may not be practice to comment it every where.

Reply:
It looks `std::select!`, but it is not, very confusing. I think comment the first one is enough.